### DISCUSSION TITLE
arbitrary nested sections using `end` (good bye sub-section)

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#381] [Ideas & RFCs]
Status: Being implemented as part of [end based argument continuation](https://github.com/FifthTry/ftd/discussions/388),

Currently we have `-- container:` to explicitly set the container after an open container.

```ftd
-- ftd.row:

-- ftd.text: first child of row
-- ftd.text: second child of row

-- container: pop
```

What if we explicitly had to demarcate any open container with `open` keyword, and it accepted a `till` argument, and we had the corresponding till ending marker.

```ftd
-- ftd.row:
open-till: the-row

-- ftd.text: first child of row
-- ftd.text: second child of row

-- end: the-row
```

Advantages:

1. we can have any number of open components, by using `--end:`
2. no need to know current parent container to user `container` properly (not true, we are planning [`container: pop`](https://github.com/FifthTry/ftd/discussions/321))

**Comment by amitu** [comment]
What if we used indentation?

```ftd
-- ftd.row:

  -- ftd.text: first child of row
  -- ftd.text: second child of row
```

Continuation syntax with `end`:

```ftd
-- foo:
warning:
> -- ftd.row:
> open-till: end
>
> -- ftd.text: left
> -- ftd.text: right
>
> -- end:
>
> -- code:
> lang: py
> 
> print("warning")

body of foo
```

Continuation syntax with indentation:

```ftd
-- foo:
warning:
> -- ftd.row:
>
>   -- ftd.text: left
>   -- ftd.text: right
>
> -- code:
> lang: py
> 
> print("warning")

body of foo
```

**Reply by amitu** [reply]
With multiline (more realistic data):

```ftd

-- ftd.row:
open-till: the-row

-- ds.markdown: 

This is a multi line text. Notice writing text here does not require
any indentation.

-- ds.h2: Some heading

More text. More multiline text. Writing things is the a key consideration
and writing indented text without indentation aware editor is pain for
writers.

-- end: the-row
```

Indentation syntax:

```ftd
-- ftd.row:

  -- ds.markdown: 

  This is a multi line text. Notice writing text here does not require
  any indentation.

  -- ds.h2: Some heading

  More text. More multiline text. Writing things is the a key consideration
  and writing indented text without indentation aware editor is pain for
  writers.
```

With 4 spaces for indentation:

```ftd
-- ftd.row:

    -- ds.markdown: 

    This is a multi line text. Notice writing text here does not require
    any indentation.

    -- ds.h2: Some heading

    More text. More multiline text. Writing things is the a key consideration
    and writing indented text without indentation aware editor is pain for
    writers.

-- ftd.row:
```

**Reply by AviralVerma13** [reply]
Since we've been against using indentations from the very start, I don't think using it is coherent with our goals for non-programmers.

**Reply by gsalunke** [reply]
@AviralVerma13 totally agreed. But we are putting it out there as a possibility so we can see how much better alternative could be if we relaxed our "never using indentation" rule (I do agree with the rule, but I am trying to see how much are we "paying for" when we are sticking with the rule).

**Reply by AviralVerma13** [reply]
I get your point. but for this, we'll have to change the parser (as right now it checks the start of section with "**--**"). Plus, our editor will have to include the indentation feature as well as  using ***tab*** for indentation to make it easy to use (like python editors).

**Reply by Heulitig** [reply]
We do not even need `open-till` line in a lot of cases if we allow the `end` to contain the same thing that "opened":

```
-- ftd.row:

-- ds.markdown: 

This is a multi line text. Notice writing text here does not require
any indentation.

-- ds.h2: Some heading

More text. More multiline text. Writing things is the a key consideration
and writing indented text without indentation aware editor is pain for
writers.

-- end: ftd.row
```

Since here `ftd.row` was open, `end: ftd.row` will close it.

If we had nested `ftd.row`s, each end will close the nearest open row.

```ftd
-- ftd.row:
-- ftd.row:
-- ftd.text: i am in inner row
-- end: ftd.row
-- ftd.text: i am in outer row
-- end: ftd.row
-- ftd.text: iam outisde the outer row
```

We have used `ftd.row`, but it can be any `-- <x>:`, it ends with corresponding `-- end: <x>`.

**Reply by Heulitig** [reply]
Also instead of `open-till`, can't we use `id`? `id` is already there to uniquely identify a container, why do we need `open-till` to also identify the container?

**Reply by Heulitig** [reply]
To give an example:

```ftd
-- ftd.row:
id: the-row

-- end: the-row
```

Maybe better than `end: ftd.row` as it more uniquely identifies what we are closing. The negative of this would be out of order closing, auto closing enforces we are closing in right order, using id means we can close in any order. But this is not really a valid objection because we can always close the wrong tag. 

On second thought: why do we ever need argument to `end`? It should just close the nearest open container? Unless we want to close several stuff? 

I think end should not take argument, and but argument should be allowed just as a hint to users, and should not have any meaning. Or maybe we can even check if the value of end is correct, even though ftd knows what it must refer to, so the identifier is added for ftd's benefit but for benefit of readers.

**Comment by amitu** [comment]
What if there was a totally different syntax, start and end to group children. This allows us to not need continuation syntax for passing arguments.

```
-- foo:
open-till: end

body of foo

-- ft.code:
lang: py

print("hello world")

-- end:

-- start: foo:warning

this is whatever

-- ft.code:
lang: py

print("Warning code")

-- end: foo:warning

**Reply by amitu** [reply]
Not fond of this. Continuation syntax with end looks better to me.

**Comment by gsalunke** [comment]
We can also use this for component definition:

```
-- ftd.column key-value-table:
caption title:
kvt list data: 
padding-top: 24
padding-bottom: 12
padding-horizontal: 24
border-width: 2
border-color: $fpm.color.main.border
border-radius: 12
background-color: $fpm.color.main.background.base
color: $fpm.color.main.text
width: auto
min-width: percent 50
spacing: 20

-- ftd.text: $title
role: $fpm.type.copy-tight
style: bold
width: fill

-- ftd.column:
width: fill

-- kv-loop:
key: $obj.key
value: $obj.value
body: $obj.body
$loop$: $data as $obj

-- end: key-value-table
```

The name of component can become the `end` "marker".

**Comment by gsalunke** [comment]
Continuation syntax more example:

```ftd
-- dark-slides.left-right-slide:
left:
> -- ds.h1: title of left stuff
>
> some left stuff
>
> -- ds.code:
> lang: py
>
> code_on("left")
right:
> -- ds.h1: title of right stuff
> heading-separator:
>> -- bling.the-end-mark:
>> end-text: Thank You
>
> some right stuff
>
> -- ds.image:
> src: $assets.right.png
```

**Reply by gsalunke** [reply]
We will still need explicit container management for handling external children.

```ftd
-- ftd.column triangular:
caption title:
boolean make-odd: false
open: true
append-at: main-container

-- ftd.row:
if: not $make-odd
anchor: parent

-- container: ftd.main

-- ftd.text: $title
if: $title is not null
color: $fpm.color.main.text
role: $fpm.type.heading-medium

-- ftd.column:
id: main-container

-- container: ftd.main

-- ftd.row:
if: not $make-odd
anchor: parent

-- end: triangular
```

Unless we get something like this to work:

```ftd
-- ftd.column triangular:
caption title:
boolean make-odd: false
open: true

-- ftd.row:
if: not $make-odd
anchor: parent
open-till: inner-row

-- ftd.text: hello

-- end: inner-row

-- ftd.text: $title
if: $title is not null

-- ftd.column:
children: $children
open: false

-- ftd.row:
if: not $make-odd
anchor: parent

-- end: triangular
```

On second thought maybe we do not need `-- container:` command at all with this syntax.

**Reply by gsalunke** [reply]
If we [remove `append-at` in favour of `$children`](https://github.com/FifthTry/ftd/discussions/380), and `open-till`/`end`, we do not need `id` inside component definition at all.

**Comment by Arpita-Jaiswal** [comment]
- Most use of the `container` is now gone.
- The only remaining use of `container` is if you want to place children inside `header`, `footer` etc.
- We can make indentation optional.

There would be two ways two close a container `end` and `end-till`.

# `end`:

The `end` closes the closest container. For instance:

```ftd
-- ftd.row: ;; row#1
  -- ftd.row: ;; row#2
  -- end: ;; closes row#2
-- end: ;; closes row#1
```

# `end-till`:

The `end-till` closes all the containers up to the container with `id` or `component name`.

- `end-till` for `id`

```ftd
-- ftd.row:
id: row-1

  -- ftd.row: 

-- end-till: row-1
```

- `end-till` for `component name`

Ends the nearest component with `component name`.

```ftd
-- ftd.row:

  -- ftd.row: 
  -- end-till: ftd.row
```

```ftd
-- ftd.row foo:

  -- ftd.row: 

-- end-till: foo
```

**Comment by amitu** [comment]
```rs
pub struct Section {
    pub name: String,
    pub caption: Option<Header>,
    pub headers: Vec<Header>,
    pub body: Option<(usize, String)>,
    pub sub_sections: Vec<Section>,
    pub is_commented: bool,
    pub line_number: usize,
}

 // in case of caption, key would be "$caption$"
enum Header {
    KV { line_number: usize, key: String, value: String },
    Section { line_number: usize, key: String, section: Section }
}
```

**Reply by amitu** [reply]
Can caption contain a section? We have some cases where we want user to mostly enter a text, but they can also pass a UI. Let's not support this. If you can pass just a string, use the till end of line and continuation syntax:

```ftd
-- foo: hello

-- foo:
> -- ftd.text: hello
```

?

**Reply by amitu** [reply]
Maybe for string continuation we do not allow:

```
-- foo: hello
> this is a long string
```

And only allow:

```
-- foo: 
> hello
> this is a long string
```

**Reply by amitu** [reply]
Can ftd component handle either string or ftd.ui as any header? We can create a wrapper:

```ftd
-- ftd.row foo:
caption ftd.ui f with wrapper ftd.text:

;; in this case we wrap hello with `ftd.text`, the "wrapper"
-- foo: hello

-- foo:
> -- some-other-text: hello
```

**Comment by Arpita-Jaiswal** [comment]
# Gains 🤩 and losses 😭

So much investment, what's the return? 

## Gains

- Indentation lovers will be able to use indentation.
- To some the new syntax looks "cleaner"
- No need to explain [`container management`](https://ftd.dev/container-management/), our most difficult concept.
  - more complex UI can be expressed with lesser mental contortion
- Less use of `id` which is needed for `-- container:` for the current component.
  - most uses of `-- container:` would be gone
- Removes to concept of sub-section. No `--- <something>:` 

## Losses

- We are adding indentation support (more like C than Python, indentation allowed but not required).  To make code clearer.
- Somewhat similar to what we have right now. Currently, `id` helps in placing items (It kind of closes all the inner containers and start placing items in the container with the given `id`). With these changes, `id` helps in closing the container with the `id` and start placing items in the parent container if any.
- Not sure if we are gaining enough with respect to user usability and experience. Suggest showing current and proposed discussion to some third person.
- So many changes in the `ftd` code.

**Comment by Heulitig** [comment]
What if there is a totally different syntax to enforce whether the component is a child or a sibling of the current container. No indentation and no sub-section

```markdown
-- ftd.row:

c-- ftd.text: first child of row
c-- ftd.text: second child of row
```

```markdown
;; We could enforce any container to avoid child behaviour
;; and act as proper siblings of the current container
-- container_1: 

s-- container_2: first sibling of container_1
s-- container_3: second sibing of container_1

;; Like we currently have this behaviour for non-container components
-- ftd.row:

-- ftd.text: first sibling of row
-- ftd.text: second sibing of row
```

We can use `close` or `close-till` to force close any open components

# close: `<no-caption>`
`close:` will close the nearest open container. It wont take any caption.
```markdown
-- ftd.row:

c-- ftd.column: 
-- close: ;; this will close the ftd.column
```

# close-till: `<open-container>`
`close-till` will either take the entire section name or an id of an open container. Using this will close all open containers till the container specified. If no caption is passed it will act same as `close:`
```markdown
-- ftd.row:

c-- ftd.column: 
-- close-till: ftd.row ;; this will close all open containers till ftd.row
```

**Comment by AbrarNitk** [comment]
# If We Implemented `end`, Why do we need `container`?

We would need it to insert stuff into other container than the main one.

```ftd
-- ds.page: Hello World

The body of the page

-- container: header-extra-links

-- ftd.text: Github
link: https://github.com/FifthTry/ftd
```

**Comment by Arpita-Jaiswal** [comment]
# Container new syntax

Since we need `container` as discussed [here](https://github.com/FifthTry/ftd/discussions/381#discussioncomment-3488120). I think we can support it for all `id`s besides external `id`s. We can do something like this:


```ftd
-- ds.page:
id: page-id

    -- ftd.row:
    id: r1
        -- ftd.row:
        id: r2
            -- something:
    -- end-till: r1

    ;; gets inside `ftd.row` type component
    -- ftd.row:
    $container$: r2
        -- ftd.text: Yo
    -- end:

    ;; gets inside `header` component in `page`
    -- ftd.row:
    $container$: page-id.header-extra-links
        -- ftd.text: Yo
    -- end:

-- end:
```

We can replace `-- container:` section with `$container$` header. 

## What we gain here:

1. This helps to close the container component (`ftd.row`, `ftd.column` etc) which is to be placed inside the container with `id` provided in `$container$` header. For instance:
```ftd
-- ftd.row:
$container$: page-id.header-extra-links
   -- ftd.text: Yo
-- end:
```
2.  We can support `$container$` header for all `id`s, not just external.

**Conclusion**: Probably get the best of both worlds here ✌️.

**Comment by amitu** [comment]
Arpita: How would `p1` parser know what is `-- end` closing?

We have three scenarios:

1. Closing by `component name`:

```ftd
-- ftd.row:

-- ftd.text: hello

-- end: ftd.row
```

Here `p1` will look for `-- <x>:` when it encounters `-- end: <x>`.

2. Closing by component id:

```ftd
-- ftd.row:
id: the-row

-- ftd.row:

-- ftd.text: hello

-- end: ftd.row

-- end: the-id
```

Here `p1` looks for `id: <x>` when it finds `-- end: <x>`.

3. Closing a component definition:

```ftd
-- ftd.row foo:

-- ftd.text: hello

-- end: foo
```

Here `p1` looks for `-- <y> <x>:` when it finds `-- end: <x>`.

4. Closing "container"s. 

```ftd
-- ds.page: Hello World

This is the page

-- container: doc-site-header

-- ftd.text: hello

-- end: doc-site-header
```

Actually we will not support `container` after this, we will use the `$container$` proposal.

# Notes

`p1` will look for all three patterns, and "close" the first one that matches. This will only work if we do not have `end-till` semantics. 

`p1` will have to know about `id`.

`-- end:` will always take parameter.

Note: These examples show components, but some of these can happen for records also.

**Reply by amitu** [reply]
In this example:

```ftd
-- ftd.row:
id: the-row

-- ftd.row:

-- ftd.text: hello

-- end: ftd.row

-- end: the-id
```

If you forget the `-- end: ftd.row` line, as per p1 parser, the inner `ftd.row` and `ftd.text` will be siblings. `p2` will have to check the fact that the `ftd.row` was not closed and should reject the program with appropriate error.

**Reply by amitu** [reply]
Arpita: If we are closing all open nodes, why do we need id? 

Valid. Probably we do not need to support "closing by id".

**Reply by amitu** [reply]
@Arpita-Jaiswal we will need a field `.closed: bool` on `Section` to indicate if a section was closed in the document. If an open container is not closed it is an error. Interpreter/Executor will have to do this check.

**Comment by amitu** [comment]
So far we have seen that a section either has single "word" section name, or it has multi words. Eg `-- foo:` or `-- integer x`. In multi word case the last word is special, it is the name, and earlier words describe the type. Should we model Section as:

```rust
pub struct Section {
    pub name: String,          // name here, eg `x` in second example
    pub kind: Option<String>,  // we can store the kind here, eg `integer`
    pub caption: Option<Header>,
    pub headers: Vec<Header>,
    pub body: Option<(usize, String)>,
    pub sub_sections: Vec<Section>,
    pub is_commented: bool,
    pub line_number: usize,
}
```

**Reply by Arpita-Jaiswal** [reply]
Looks good 😊
